### DISCUSSION TITLE
docs: Missing best practice for Seeding when deep nested relationships

### DISCUSSION THREAD

**Discussion by tobiaskraus** [mikro-orm/mikro-orm#2964] [Q&A] [answered]
## Problem

We have a database schema where around 100 tables are almost all related to each other. We want to use `@mikro-orm/seeder` for test purposes. The idea is to have one Seeder file per entity. The documentation explains only in a small paragraph about [Factory relationships](https://mikro-orm.io/docs/next/seeding#factory-relationships):

> ManyToOne and OneToOne relations
>
> ```ts
> const books: Book[] = new BookFactory(orm.em).each(book => {
>     book.author = new AuthorFactory(orm.em).makeOne();
> }).make(5);
> ```

> OneToMany and ManyToMany
> 
> ```ts
> const books: Book[] = new BookFactory(orm.em).each(book => {
>     book.owners.set(new OwnerFactory(orm.em).make(5));
> }).make(5);
>  ```

But this is super unpractical for complex databases, as I might want to create a Book with special values and use this Book in multiple other Many relations (Owner, Sold, BookCopy, ...). And I would like to avoid having one file with multiple nested `.each`.
I think it's especially confusing to do it in this way, if some Many-relations have a relation by themselves with another Many-relation (like here `baz`) where I want to re-use the same instances:

```ts
const baz = new BazFactory(orm.em).make(2);

const foos: Foo[] = new FooFactory(orm.em).each(foo => {
    foo.bars.set(new BarFactory(orm.em).each(bar =>
        bar.bazs.set(baz);
    ).make(5));
     foo.qux.set(new QuxFactory(orm.em).each(qux =>
        qux.bazs.set(baz);
    ).make(5));
}).make(5);
```

```
    foo
   /   \
bars   qux
   \   /
    baz 
```

I'm sorry if this is more a question, then a `docu` suggestion, but other people might have the same issues when reading the docs about seeding.

**Comment by B4nan** [comment]
I am not sure what kind of advice you would like to get here, if you want complex entity graph, you will need to build it yourself. I am not against having some native helpers for it, but let's first define the expectations of such thing. 

cc @Langstra

**Comment by Langstra** [comment]
To be honest I am not sure how this could be done best. Maybe if you want to reuse existing instances you should seed them first. Then when you want to attach them to different types of entities you can attach them in later loops.
```
const bazs = new BazFactory(orm.em).make(2);
const foos = new FooFactory(orm.em).make(2);
const qux = new QuxFactory(orm.em).make(2);
const bars = new BarFactory(orm.em).make(2);

bazs.forEach(baz => baz.bars.set(bars));
// etc

orm.em.flush();
```
Have not tried this, so I am not sure about this. Just thinking out loud here ;)

**Comment by tobiaskraus** [comment, marked as answer]
Thank you both for your fast reply.

We currently chose this way - and it seems to be fine for us:

- one `DatabaseSeeder.ts`
- each entity has its own factory
- starting with instances which have no OneToMany relationships to other entities (the lose ends)
- mostly we added a `makeAll(...params)` method in the factory to generate specific instances of its entity and gets required owner relationships as params passed to it

```ts
// DatabaseSeeder.ts:
const company = companyFactory.makeOne();
const users = userFactory.make(2, { companyId: company.id });
const buildings = buildingFactory.makeAll(users);
const roofs = roofFactory.makeAll(buildings[0], users);
// ...

// RoofFactory.ts
makeAll(building: Building, users: User[]): Roof[] {
  return [
      this.makeOne({
        buildingId: building.id,
        lastUpdatedBy: users[0],
        // ...
      }),
      this.makeOne({ buildingId: building.id }),
      this.makeOne({ name: "deleted Roof", buildingId: building.id, deletedAt: new Date() }),
  ];
}
```

So in the end we didn't use the Seeders `call` method to call other seeders described in documentation under ["Calling additional seeders"](https://mikro-orm.io/docs/next/seeding#calling-additional-seeders). But it works just fine, with only one Seeder file and implementing everything data specific in the factories.
